lib/rbi/rbs/type_translator.rb

```diff
@@ -0,0 +1,164 @@
+            ::RBS::Types::Optional,
+            ::RBS::Types::Proc,
+            ::RBS::Types::Record,
+            ::RBS::Types::Tuple,
+            ::RBS::Types::Union,
+            ::RBS::Types::UntypedFunction,
+            ::RBS::Types::Variable,
+          )
+        end
+
+        #: (NodeType) -> Type
+        def translate(type)
+          case type
+          when ::RBS::Types::Alias
+            # TODO: unsupported yet
+            Type.untyped
+          when ::RBS::Types::Bases::Any
+            Type.untyped
+          when ::RBS::Types::Bases::Bool
+            Type.boolean
+          when ::RBS::Types::Bases::Bottom
+            Type.noreturn
+          when ::RBS::Types::Bases::Class
+            # TODO: unsupported yet
+            Type.untyped
+          when ::RBS::Types::Bases::Instance
+            Type.attached_class
+          when ::RBS::Types::Bases::Nil
+            Type.simple("NilClass")
+          when ::RBS::Types::Bases::Self
+            Type.self_type
+          when ::RBS::Types::Bases::Top
+            Type.anything
+          when ::RBS::Types::Bases::Void
+            Type.void
+          when ::RBS::Types::ClassSingleton
+            Type.class_of(Type.simple(translate_name(type.name)))
+          when ::RBS::Types::ClassInstance
+            Type.simple(translate_name(type.name))
+          when ::RBS::Types::Function
+            translate_function(type)
+          when ::RBS::Types::Interface
+            # TODO: unsupported yet
+            Type.untyped
+          when ::RBS::Types::Intersection
+            Type.all(*type.types.map { |t| translate(t) })
+          when ::RBS::Types::Literal
+            # TODO: unsupported yet
+            Type.untyped
+          when ::RBS::Types::Optional
+            Type.nilable(translate(type.type))
+          when ::RBS::Types::Proc
+            proc = T.cast(translate(type.type), Type::Proc)
+            proc.bind(translate(type.self_type)) if type.self_type
+            proc
+          when ::RBS::Types::Record
+            Type.shape(type.fields.map { |name, type| [name, translate(type)] }.to_h)
+          when ::RBS::Types::Tuple
+            Type.tuple(type.types.map { |t| translate(t) })
+          when ::RBS::Types::Union
+            Type.any(*type.types.map { |t| translate(t) })
+          when ::RBS::Types::UntypedFunction
+            Type.proc.params(arg0: Type.untyped).returns(Type.untyped)
+          when ::RBS::Types::Variable
+            Type.type_parameter(type.name)
+          else
+            T.absurd(type)
+          end
+        end
+
+        private
+
+        #: (::RBS::TypeName) -> String
+        def translate_name(name)
+          string = String.new
+
+          namespace = name.namespace
+          string << "::" if namespace.absolute?
+
+          path = namespace.path
+          string << path.join("::")
+          string << "::" if path.any?
+
+          string << name.name.to_s
+
+          string
```

Comment:
I have a few points about this:
1. Let's not use `String.new` for creating a mutable string, since it has a footgun: It always created an `ASCII-8BIT` string, despite the encoding of the Ruby file. Instead, we should use `"".dup` or `+""` to create mutable strings, and
2. I believe this is exactly what `RBS::TypeName#to_s` does internally anyway, so I am not sure why we need this at all. Ref: https://github.com/ruby/rbs/blob/master/lib/rbs/type_name.rb#L35-L37 and https://github.com/ruby/rbs/blob/master/lib/rbs/namespace.rb#L75-L82